app/components/top-nav.tsx

```diff
@@ -1,4 +1,11 @@
+import { useMatches } from '@remix-run/react';
+
+import type { UserSession } from '~/modules/session/session.server';
+
 export default function TopNav() {
+  const matches = useMatches();
```

Comment:
Great job looking into Remix for a better solution than prop drilling and React Context!

Remix also added a new `useRouteLoaderData` hook ([see docs](https://remix.run/docs/en/main/hooks/use-route-loader-data)) for convenience. `useRouteLoaderData` probably uses `useMatches` under the hood, so it's just a nice layer of abstraction!

In general, with `useMatches`, I would not trust the order in the array and instead try to find the right loader data by looking for the route id (`root` in this case).

Finally, maybe we can abstract the `useRouteData('root')` call in a reuseable hook like `useUser` or `useIsLoggedIn()`?

Reply:
`useRouteLoaderData` seems perfect for this! I'll go ahead and make a custom hook using this method. Thanks for the tip!